plutus-tx/src/PlutusTx/DataMap.hs

```diff
@@ -0,0 +1,355 @@
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts   #-}
+{-# LANGUAGE LambdaCase         #-}
+{-# LANGUAGE NoImplicitPrelude  #-}
+{-# LANGUAGE TupleSections      #-}
+{-# LANGUAGE ViewPatterns       #-}
+
+module PlutusTx.DataMap where
+
+import PlutusTx.Builtins qualified as P
+import PlutusTx.Builtins.Internal qualified as BI
+import PlutusTx.IsData qualified as P
+import PlutusTx.Prelude hiding (null, toList)
+import PlutusTx.These
+
+import Prelude qualified as Haskell
+
+{- | An associative map implementation backed by `P.BuiltinData`.
+
+This map implementation has the following characteristics:
+
+  * The `P.toBuiltinData` and `P.unsafeFromBuiltinData` operations are no-op.
+  * Other operations are slower than @PlutusTx.AssocMap.Map@, although equality
+    checks on keys can be faster due to `P.equalsData`.
+  * Many operations involve converting the keys and/or values to/from `P.BuiltinData`.
+
+Therefore this map implementation is likely a better choice than @PlutusTx.AssocMap.Map@
+if it is part of a data type defined using @asData@, and the key and value types
+have efficient `P.toBuiltinData` and `P.unsafeFromBuiltinData` operations (e.g., they
+are primitive types or types defined using @asData@).
+-}
+newtype Map k a = Map P.BuiltinData
+  deriving stock (Haskell.Eq, Haskell.Show)
+  deriving newtype (Eq)
+
+instance P.ToData (Map k a) where
+  {-# INLINEABLE toBuiltinData #-}
+  toBuiltinData (Map d) = d
+
+instance P.FromData (Map k a) where
+  fromBuiltinData = Just . Map
+
+instance P.UnsafeFromData (Map k a) where
+  unsafeFromBuiltinData = Map
+
+{-# INLINEABLE lookup #-}
+lookup :: forall k a. (P.ToData k, P.UnsafeFromData a) => k -> Map k a -> Maybe a
+lookup (P.toBuiltinData -> k) m = case lookup' k (toBuiltinList m) of
+  Just a  -> Just (P.unsafeFromBuiltinData a)
+  Nothing -> Nothing
+
+{-# INLINEABLE lookup' #-}
+lookup' ::
+  BuiltinData ->
+  BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) ->
+  Maybe BuiltinData
+lookup' k = go
+  where
+    go xs =
+      P.matchList
+        xs
+        Nothing
+        ( \hd tl ->
+            let k' = BI.fst hd
```

Comment:
The alignment seems a bit odd here (and elsewhere in this file).

Reply:
It's fourmolu - it would take me forever to hand-format this module!

Reply:
> It's fourmolu - it would take me forever to hand-format this module!

Ahhhh.  I've never tried using fourmolu.  I just let stylish-haskell do its thing, and it does something totally different!